src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala

```diff
@@ -2,50 +2,58 @@ package mesosphere.marathon
+  * scale of a service when instances become UnreachableInactive, or come back
+  * from that state into one that is considered active.
+  *
+  * The result will be a ScaleRunSpec directive. This will be ignored during
+  * ongoing deployments for the service. In this case, the responsible
+  * deployment logic will have to handle these cases.
   */
 class ScaleAppUpdateStepImpl @Inject() (
     @Named("schedulerActor") schedulerActorProvider: Provider[ActorRef]) extends InstanceChangeHandler with StrictLogging {
 
   private[this] lazy val schedulerActor = schedulerActorProvider.get()
 
-  private[this] def scalingWorthy: Condition => Boolean = {
-    case Condition.Reserved | Condition.UnreachableInactive | _: Condition.Terminal => true
-    case _ => false
-  }
-
   override def name: String = "scaleApp"
   override def metricName: String = "scale-app"
 
   override def process(update: InstanceChange): Future[Done] = {
-    // TODO(PODS): it should be up to a tbd TaskUnreachableBehavior how to handle Unreachable
-    calcScaleEvent(update).foreach(event => schedulerActor ! event)
+    // This doesn't block on the result of a scale check, which means that
+    // fast subsequent instance changes for the same PathId will eventually be
+    // ignored, since the pathId might be locked from a previous ScaleRunSpec
+    // attempt. This happened in tests, so it should be considered a real problem.
+    // 1. We would send an event to the schedulerActor
+    // 2. The schedulerActor would lock the app and call schedulerActions.scale
+    // 3. If one or more InstanceChange are processed before the lock is released, no replacements are scheduled for those
+    // The logs say e.g. `Did not try to scale run spec /app-kill-all-tasks-of-an-app; it is locked`
+    // An option would be to block here until the scale request is acked.
+    update.lastState.foreach { lastState =>
+      calcScaleEvent(update.id, lastState.condition, update.condition).foreach(event => schedulerActor ! event)
+    }
     Future.successful(Done)
   }
 
-  def calcScaleEvent(update: InstanceChange): Option[ScaleRunSpec] = {
-    if (scalingWorthy(update.condition) && update.lastState.forall(lastState => !scalingWorthy(lastState.condition))) {
-
-      val runSpecId = update.runSpecId
-      val instanceId = update.id
-      val state = update.condition
-      logger.info(s"initiating a scale check for runSpec [$runSpecId] due to [$instanceId] $state")
-      // TODO(PODS): we should rename the Message and make the SchedulerActor generic
-      Some(ScaleRunSpec(runSpecId))
+  def calcScaleEvent(instanceId: Instance.Id, lastCondition: Condition, newCondition: Condition): Option[ScaleRunSpec] = {
+    if (lastCondition == Condition.UnreachableInactive || newCondition == Condition.UnreachableInactive) {
+      // if the instance was UnreachableInactive or just became UnreachableInactive, we need to adjust the scale
```

Comment:
If I'm not mistaken this is done somewhere else. I have to check how we handled unreachable instances before.

Ideally an orchestrator will handle unreachable instances. There is a ticket for that.